solarforecastarbiter/reports/main.py

```diff
@@ -0,0 +1,273 @@
+
+
+def create_metadata(report):
+    """
+    Create prereport metadata.
+
+    Returns
+    -------
+    metadata: dict
+    """
+    metadata = dict(
+        name=report.name, start=report.start, end=report.end,
+        now=pd.Timestamp.utcnow())
+    metadata['versions'] = get_versions()
+    return metadata
+
+
+def get_versions():
+    packages = [
+        'solarforecastarbiter',
+        'pvlib',
+        'pandas',
+        'numpy',
+        'bokeh',
+        'netcdf4',
+        'xarray',
+        'tables',
+        'numexpr',
+        'bottleneck',
+        'jinja2',
+    ]
+    versions = {}
+    for p in packages:
+        try:
+            v = pkg_resources.get_distribution(p).version
+        except DistributionNotFound:
+            v = 'None'
+        versions[p] = v
+    versions['python'] = platform.python_version()
+    versions['platform'] = platform.platform()
+    return versions
+
+
+def get_data_for_report_embed(session, report):
+    """
+    Get time series data for report.
+
+    1 API call is made for each unique forecast and observation object.
+
+    Parameters
+    ----------
+    session : solarforecastarbiter.api.APISession
+        API session for getting and posting data
+    report : solarforecastarbiter.datamodel.Report
+        Metadata describing report
+
+    Returns
+    -------
+    fx_obs_cds : list
+        List of (forecast, observation, ColumnDataSource) tuples to
+        pass to bokeh plotting objects.
+    """
+    data = get_data_for_report(session, report)
+    fxobs_data = data_dict_to_fxobs_data(data, report)
+    fx_obs_cds = [figures.construct_fx_obs_cds(k, v[0], v[1])
+                  for k, v in fxobs_data.items()]
+    return fx_obs_cds
+
+
+def create_prereport_from_data(report, data):
+    """
+    Create a pre-report using data and report metadata.
+
+    The prereport is a markdown file with all elements rendered except
+    for bokeh plots.
+
+    Parameters
+    ----------
+    report : solarforecastarbiter.datamodel.Report
+        Metadata describing report
+    data : dict
+        Keys are all Forecast and Observation objects in the report,
+        values are the corresponding data.
+
+    Returns
+    -------
+    metadata : str
+        prereport metadata in JSON format.
+    prereport : str
+        prereport in markdown format.
+    """
+    # call function: metrics.align_observations_forecasts
+    # call function: metrics.calculate_many
+    # call function: reports.metrics_to_JSON
+    # call function: add some metadata to JSON
+    # call function: configure tables and figures, add to JSON
+    # call function: pre-render report in md format
+    # return json, prereport
+
+    # debug
+    from solarforecastarbiter.reports.tests.test_main import dummy_metrics
+    metrics = dummy_metrics.copy()
+
+    metadata = create_metadata(report)
+    prereport = template.prereport(metadata, metrics)
+
+    # put the metrics in the metadata because why not
```

Comment:
because size, so we should make sure it's reasonable

Reply:
They need to go somewhere and this is the only place they're currently saved. Should the eventual post to the API include metadata, metrics, and prereport or metadata+metrics and prereport?

Reply:
still not sure. fine to keep it here as long as it is easy to swap around later